clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -723,73 +718,60 @@ public static Sensor throttleTimeSensor(Metrics metrics, SenderMetricsRegistry m
         public final Sensor compressionRateSensor;
         public final Sensor maxRecordSizeSensor;
         public final Sensor batchSplitSensor;
-        private SenderMetricsRegistry metricsRegistry;
+        private final SenderMetricsRegistry metrics;
 
-        public SenderMetrics(Metrics metrics, SenderMetricsRegistry metricsRegistry) {
+        public SenderMetrics(SenderMetricsRegistry metrics) {
             this.metrics = metrics;
-            this.metricsRegistry = metricsRegistry;
 
             this.batchSizeSensor = metrics.sensor("batch-size");
-            MetricName m = metrics.metricInstance(metricsRegistry.batchSizeAvg);
-            this.batchSizeSensor.add(m, new Avg());
-            m = metrics.metricInstance(metricsRegistry.batchSizeMax);
-            this.batchSizeSensor.add(m, new Max());
+            this.batchSizeSensor.add(metrics.batchSizeAvg, new Avg());
```

Comment:
Look at how clean and concise it is now. Direct variable access, and no need to call any methods, no need to use the Metrics object, and nothing is exposed about the metricsInstance or MetricNameTemplates.